src/gpu/amd/miopen_matmul.hpp

```diff
@@ -0,0 +1,176 @@
+/*******************************************************************************
+* Copyright 2020-2022 Intel Corporation
+* Copyright 2020 Codeplay Software Limited
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef GPU_AMD_MIOPEN_MATMUL_HPP
+#define GPU_AMD_MIOPEN_MATMUL_HPP
+
+#include <assert.h>
+
+#include "common/matmul_pd.hpp"
+#include "common/primitive.hpp"
+
+#include "gpu/amd/miopen_matmul_executor.hpp"
+#include "gpu/amd/miopen_matmul_impl.hpp"
+#include "gpu/amd/sycl_hip_utils.hpp"
+
+namespace dnnl {
+namespace impl {
+namespace gpu {
+namespace amd {
+
+struct miopen_matmul_t : public primitive_t {
+    using primitive_t::primitive_t;
+    struct pd_t : public matmul_pd_t {
+        using matmul_pd_t::matmul_pd_t;
+
+        DECLARE_COMMON_PD_T("hip:miopen:any", miopen_matmul_t);
+
+        status_t init(engine_t *) {
+            using namespace data_type;
+            using smask_t = primitive_attr_t::skip_mask_t;
+
+            data_type_t src_dt = src_md()->data_type;
+            data_type_t dst_dt = dst_md()->data_type;
+            data_type_t wei_dt = weights_md(0)->data_type;
+            data_type_t bia_dt
+                    = with_bias() ? weights_md(1)->data_type : data_type::f32;
+
+            bool f32_case = utils::everyone_is(f32, src_dt, wei_dt, dst_dt);
+            bool f16_case = utils::everyone_is(f16, src_dt, wei_dt, dst_dt);
+            bool s8_case
+                    = utils::everyone_is(s8, src_dt, wei_dt) && (dst_dt == s32);
+            bool bf16_case = utils::everyone_is(bf16, src_dt, wei_dt)
```

Comment:
Given that `bf16` support is not widely available in AMD GPUs I think we need to check if the device supports it. You can probably implement a function that would inspect the device to check whether it supports `bf16` or not. I think the following HIP API can provide you necessary information: `hipError_t hipGetDeviceProperties (hipDeviceProp_t ∗prop, int deviceId)`.

Reply:
I think you can add the function below to `sycl_hip_utils.cpp`. This will return `true` if device supports bf16 data type.
```cpp
bool has_bf16_support(const ::sycl::device &dev) {
    auto hip_dev = compat::get_native<HIPdevice>(dev);
    hipDeviceProp_t props {};
    HIP_EXECUTE_FUNC(hipGetDeviceProperties, &props, hip_dev);

    std::string arch_name(props.gcnArchName);
    arch_name = std::string(arch_name).substr(0, arch_name.find(":"));
    if (utils::everyone_is(std::string::npos, arch_name.find("gfx90a"),
                arch_name.find("gfx908"))) {
        return false;
    }
    return true;
}
```

Reply:
As per latest discussion didn't done this once since at present don't know which all GPU's will support bf16.

Reply:
Decision is to always assume that rocBLAS can handle bf16 data type on hardware that doesn't support bf16 instructions.